packages/artifact/src/upload-artifact-http-client.ts

```diff
@@ -0,0 +1,299 @@
+  const CHUNK_CONCURRENCY =
+    parseEnvNumber('ARTIFACT_CHUNK_UPLOAD_CONCURRENCY') ||
+    defaultChunkUploadConcurrency
+  const MAX_CHUNK_SIZE =
+    parseEnvNumber('ARTIFACT_UPLOAD_CHUNK_SIZE') || 4 * 1024 * 1024 // 4 MB Chunks
+  debug(
+    `File Concurrency: ${FILE_CONCURRENCY}, Chunk Concurrency: ${CHUNK_CONCURRENCY} and Chunk Size: ${MAX_CHUNK_SIZE}`
+  )
+
+  const parameters: UploadFileParameters[] = []
+
+  // Prepare the necessary parameters to upload all the files
+  for (const file of filesToUpload) {
+    const resourceUrl = new URL(uploadUrl)
+    resourceUrl.searchParams.append(
+      'scope',
+      '00000000-0000-0000-0000-000000000000'
+    )
+    resourceUrl.searchParams.append('itemPath', file.uploadFilePath)
+    parameters.push({
+      file: file.absoluteFilePath,
+      resourceUrl: resourceUrl.toString(),
+      restClient: client,
+      concurrency: CHUNK_CONCURRENCY,
+      maxChunkSize: MAX_CHUNK_SIZE
+    })
+  }
+
+  const parallelUploads = [...new Array(FILE_CONCURRENCY).keys()]
+  const fileSizes: number[] = []
+  let uploadedFiles = 0
+
+  // Only allow a certain amount of files to be uploaded at once, this is done to reduce errors if
+  // trying to upload everything at once
+  await Promise.all(
+    parallelUploads.map(async () => {
+      while (uploadedFiles < filesToUpload.length) {
+        const currentFileParameters = parameters[uploadedFiles]
+        uploadedFiles += 1
+        fileSizes.push(await uploadFileAsync(currentFileParameters))
+      }
+    })
+  )
+
+  let sum = 0
+  for (const fileSize of fileSizes) {
+    sum += fileSize
+  }
+  console.log(`Total size of all the files uploaded ${sum}`)
+  return sum
+}
+
+/**
+ * Asyncronously uploads a file. If the file is bigger than the max chunk size it will be uploaded via multiple calls
+ * @param {UploadFileParameters} parameters Information about the files that need to be uploaded
+ * @returns The size of the file that was uploaded in bytes
+ */
+async function uploadFileAsync(
+  parameters: UploadFileParameters
+): Promise<number> {
+  const fileSize: number = fs.statSync(parameters.file).size
+  const parallelUploads = [...new Array(parameters.concurrency).keys()]
+  let offset = 0
+
+  await Promise.all(
+    parallelUploads.map(async () => {
+      while (offset < fileSize) {
+        const chunkSize = Math.min(fileSize - offset, parameters.maxChunkSize)
+        const start = offset
+        const end = offset + chunkSize - 1
+        offset += parameters.maxChunkSize
+        const chunk: NodeJS.ReadableStream = fs.createReadStream(
+          parameters.file,
+          {
+            start,
+            end,
+            autoClose: false
+          }
+        )
+
+        await uploadChunk(
+          parameters.restClient,
+          parameters.resourceUrl,
+          chunk,
+          start,
+          end,
+          fileSize
+        )
+      }
+    })
+  )
+  return fs.statSync(parameters.file).size
```

Comment:
Can you re-use the file size from above here? No need to make the extra file system call:
```suggestion
  return fileSize
```